docs/Tutorial_DiscordMERLIN.ipynb

```diff
@@ -0,0 +1,1356 @@
+{
```

Comment:
<pre class="ql-syntax" spellcheck="false">Line #1.    def murlin(subseqs, min_dist, k=1, warm_up_percentage=0.01, percentage=0.05):
</pre><p>I think the input should be <code>T</code> and <code>m</code> and not <code>subseqs</code></p><p> </p><p>Instead of <code>warm_up_percentage</code> and <code>percentage</code> , can we just eliminate <code>warm_up_percentage</code> and use<code>percentage=0.01</code> everywhere instead? This way, regardless of whether it's the first iteration or the nth iteration, the decay rate is linear and the same.</p><p> </p><p>I'm just trying to remove as much of the branching <code>if/else</code>  as possible</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/505/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='40'/>

Reply:
> I think the input should be T and m and not subseqs

I know you said that `murlin` is top-level. But, I am thinking of letting users use `merlin` only as it will call the `murlin` and do the job even if the user wants to do the anomaly detection for a discord with particular length rather than range of  lengths.

(But, if you are thinking about doing the z-normalizing on the fly, you are correct. I should use `T` and `m`)



> can we just eliminate `warm_up_percentage` and use`percentage=0.01` everywhere instead?

We can. However, the idea is to have a **higher** `rate of change` when we are trying to find **`top-k (k>1)`** discords. Because, as we are excluding trivial matches of the discovered discords, we would like to reduce `min_dist` more than usual. 

If we optimize the process in the later versions, choosing the same `percentage` might be ok and have no significant impact on the computation time.

Reply:
<p>I think we need to take some time to think about if there's a better/different way to approach the problem so that we can avoid logic like:</p><p> </p><pre class="ql-syntax" spellcheck="false">if i==0:&nbsp;
    max_attempts = int(np.ceil(1/warm_up_percentage))&nbsp;
else:
&nbsp; &nbsp; max_attempts = int(np.ceil(1/percentage))
</pre><p>or</p><pre class="ql-syntax" spellcheck="false">if i==0:
    min_dist = min_dist * (1 - warm_up_percentage * n_attempts)
else:
&nbsp; &nbsp; min_dist = min_dist * (1 - percentage * n_attempts)
</pre><p>It would be great if we could simplify/minimize/eliminate the <code>if/else</code>  and still be flexible in the future.</p>

Reply:
So, I thought about it. If we want to eliminate `if/else` and at the same time try to stay close to the `MERLIN` algorithm, we can do:

```
percentage = 0.05*min(1, i) + 0.01
```
For `i=0`, we will get `0.01` and for `i>=0` we will get `0.06` as the percentage. I know it is not `0.05` but that should not be a problem. The idea is just to have a slightly(?) larger percentage for `i>0`.

I do not know how readable it is from your perspective. Because it basically does the same `if/else`. We can get rid of both `if/else` with this line of code.

What do you think?

Reply:
<p>I am thinking of something that is even simpler:</p><p> </p><ul><li>We remove/replace both <code>warm_up_percentage</code>and  <code>percentage</code>  with a single argument called <code>decay</code> , which controls the decay of <code>min_dist</code>  </li><li><code>decay</code>  has a default value of <code>None</code>  and, in that case, within the body of the function we do:</li></ul><pre class="ql-syntax ql-indent-1" spellcheck="false">decay = np.full(k, 0.05)
decay[0] = 0.01
</pre><ul><li>This then allows us to reference <code>decay[i]</code>  easily within each round of the outer <code>for-loop</code> . So, you'd then be able to do:</li></ul><pre class="ql-syntax" spellcheck="false">max_attempts = int(np.ceil(1/decay[i]))&nbsp;
</pre><p>and</p><pre class="ql-syntax" spellcheck="false">min_dist = min_dist * (1 - decay[i] * n_attempts)
</pre><p> </p><p>Of course, if the user wants to set <code>decay</code>  then they must provide a list of floats (between <code>0</code>  and <code>1</code>)  with a <code>len(decay) == k</code>  and you'll need to check for this. However, the benefit here is that we get total flexibility over how <code>min_dist</code>  is being decayed, things are more explicit, and I think it simplifies our code a lot.</p><p> </p><p>What do you think?</p>

Reply:
> the benefit here is that we get total flexibility over how min_dist is being decayed, things are more explicit, and I think it simplifies our code a lot

Right. This is a better approach!! Thanks for the suggestion.


> Of course, if the user wants to set decay then they must provide a list of floats (between 0 and 1) 

I think it might be a good idea to add a note (in docstring) saying that setting bad values for `decay` may increase the computation time.

Reply:
<p>Yes, I hope this is the easy part :)</p>